server/services/caldav/lib/calendar/calendar.formaters.js

```diff
@@ -0,0 +1,198 @@
+        external_id: `${event.uid}${startDate.format('YYYY-MM-DD')}`,
+        selector: slugify(`${recurrenceTitle} ${startDate.format('YYYY-MM-DD')}`),
+        name: recurrenceTitle,
+        location: event.location,
+        url: event.href,
+        calendar_id: gladysCalendar.id,
+      };
+
+      if (event.start && event.start.tz === undefined) {
+        newEvent.full_day = true;
+      }
+
+      if (newEvent.full_day) {
+        startDate.set({ hour: 0, minute: 0, second: 0, millisecond: 0 });
+        endDate.set({ hour: 0, minute: 0, second: 0, millisecond: 0 });
+      }
+
+      newEvent.start = startDate.toISOString();
+      newEvent.end = endDate.toISOString();
+
+      return newEvent;
+    }
+
+    return null;
+  });
+}
+
+/**
+ * @description Format events for Gladys calendar compatibility.
+ * @param {Array} caldavEvents - Events to format.
+ * @param {Object} gladysCalendar - Gladys calendar where events are saved.
+ * @returns {Array} All events formatted.
+ * @example
+ * formatEvents(caldavEvents, gladysCalendar)
+ */
+function formatEvents(caldavEvents, gladysCalendar) {
+  let events = [];
+
+  caldavEvents.forEach((caldavEvent) => {
+    if (caldavEvent.type !== 'VEVENT') {
+      return;
+    }
+
+    if (typeof caldavEvent.rrule === 'undefined') {
+      const newEvent = {
+        external_id: caldavEvent.uid,
+        selector: slugify(`${caldavEvent.summary} ${this.moment(caldavEvent.start).format('YYYY-MM-DD')}`),
+        name: caldavEvent.summary,
+        location: caldavEvent.location,
+        url: caldavEvent.href,
+        calendar_id: gladysCalendar.id,
+      };
+
+      if (caldavEvent.start) {
+        newEvent.start = caldavEvent.start.toISOString();
+      }
+
+      if (caldavEvent.end) {
+        newEvent.end = caldavEvent.end.toISOString();
+      }
+
+      if (
+        caldavEvent.start &&
+        caldavEvent.start.tz === undefined &&
+        Number.isInteger(this.moment(caldavEvent.end).diff(this.moment(caldavEvent.start), 'days', true))
+      ) {
+        newEvent.full_day = true;
+      }
+
+      events.push(newEvent);
+    } else {
+      events = events.concat(this.formatRecurringEvents(caldavEvent, gladysCalendar).filter((e) => e !== null));
+    }
+  });
+
+  return events;
+}
+
+/**
+ * @description Format calendar for Gladys compatibility.
+ * @param {Array} caldavCalendars - Dav calendars to format.
+ * @param {Object} userId - Gladys user, calendar owner.
+ * @returns {Array} Formatted calendars.
+ * @example
+ * formatCalendars(calendars, userId)
+ */
+function formatCalendars(caldavCalendars, userId) {
+  const calendars = [];
+  caldavCalendars.forEach((caldavCalendar) => {
+    const newCalendar = {
+      external_id: caldavCalendar.url,
+      name: caldavCalendar.displayName,
+      description: caldavCalendar.description || `Calendar ${caldavCalendar.displayName}`,
+      service_id: this.serviceId,
+      user_id: userId,
+      ctag: caldavCalendar.ctag,
```

Comment:
What's a ctag and why is it useful ? (just asking)

Reply:
`ctag` is like a calendar status. If `ctag` changes, it means events change inside the calendar (create/update/delete). With this, we can skip useless requests to fetch event changes.